cluster-api-resource-monitoring/README.md

```diff
@@ -0,0 +1,327 @@
+
+##### Pros
+
+- Code ownership is handled per provider, which reflects the separation of our current teams.
+- Every CAPI related CR is monitored by one single App.
+
+##### Cons
+
+- Some code maintenance must be done multiple times.
+
+#### Variant 2: one CASM instance per infrastructure provider plus one CASM instance for CAPI core
+
+##### Pros
+
+- Cross-provider related components (CAPI core) are shared via one dedicated app.
+
+##### Cons
+
+- Waste of resources as two versions of CASM are deployed in management clusters.
+
+## kube state metrics
+
+With version [`v2.5.0`](https://github.com/kubernetes/kube-state-metrics/releases/tag/v2.5.0) of `kube-state-metrics` it's now possible to create metrics from custom resources by defining a configuration per metric.
+
+As of now (`v2.5.0`) it's not possible to create metrics for non-numbered fields, e.g.:
+
+```yaml
+kind: CustomResourceStateMetrics
+spec: 
+  resources: 
+    - groupVersionKind: 
+        group: cluster.x-k8s.io
+        kind: Machine
+        version: v1beta1
+      metrics: 
+        - each: 
+            path: 
+              - status
+              - phase
+          help: "machine phase"
+          name: phase
+```
+
+will lead to following error:
+
+```
+E0607 09:12:00.168168  126676 registry_factory.go:469] "kube_cluster_x-k8s_io_v1beta1_Machine_phase" err="[status,phase]: []: strconv.ParseFloat: parsing \"Running\": invalid syntax"
+```
+
+This behavior got already addressed to Cluster API community in [issue #6458](https://github.com/kubernetes-sigs/cluster-api/issues/6458#issuecomment-1148293873). The current discussion together with the KSM folks is one in [this slack thread](https://kubernetes.slack.com/archives/CJJ529RUY/p1654593034854759).
+
+With the assumption, that `kube-state-metrics` accept patches which might make CASM obsolete, this section will cover the integration of Cluster API infrastructure provider specific `CustomResourceStateMetrics` configuration.
+
+### Variant 1: specific `kube-state-metrics` configuration for the already existing `kube-state-metrics-app`
+
+As `kube-state-metrics` is already deployed in each cluster (no matter if management- or workload cluster), a condition has to be implemented to detect, if the current KSM instance is running on a management- or a workload cluster. With this condition in place, we can provide a CAPI generic `CustomResourceStateMetrics` configuration plus one `CustomResourceStateMetrics` configuration per infrastructure provider.
+
+#### Pros
+
+- CAPI generic `CustomResourceStateMetrics` configuration could be easily re-used on different providers
+- KSM-App already exists
+
+#### Cons
+
+- Scope of KSM between management- and workload cluster differs
+  - resource requirements
+  - permissions
+- Changes in KSM-App affect all clusters, even if changes are only done for management clusters (e.g. version bump)
```

Comment:
can we just change those values at deployment time depending on the target cluster? I think we do for other components too

Reply:
My understanding is YES and this seems the best solution to me. We already use helm templates to differentiate behavior of apps (including monitoring stuff) according to the environment.

Reply:
yes, it's possible. It's more about accepting the impact when we go this way

Reply:
That's exactly one idea, KSM folks mentioned someday in slack (TODO marioc: link to that statement) that each third party component can ship there own `CustomResourceStateMetrics` configuration and `KSM` is taking care of it.

Currently `CustomResourceStateMetrics` is only defined as key in the configmap, which `KSM` will read.